lib/password.c

```diff
@@ -1,32 +1,36 @@
 #define BITLBEE_CORE
+
+#include <gcrypt.h>
 #include "bitlbee.h"
 #include "base64.h"
 #include "arc.h"
 #include "md5.h"
 
+
+/* pbkdf2-hashed password. We store an algoritm identifier ("1"), the salt
+ * used, the number of rounds and the final hash */
 size_t password_hash(const char *password, char **hash)
 {
-	md5_byte_t pass_md5[21];
-	md5_state_t md5_state;
+	unsigned char rawhash[64], salt[8];
+	char *b64hash, *b64salt;
+	int saltlen = 8, rounds = 10000, keylen = 64;
 
-	/* Generate a salted md5sum of the password. Use 5 bytes for the salt
-	   (to prevent dictionary lookups of passwords) to end up with a 21-
-	   byte password hash, more convenient for base64 encoding. */
-	random_bytes(pass_md5 + 16, 5);
-	md5_init(&md5_state);
-	md5_append(&md5_state, (md5_byte_t *) password, strlen(password));
-	md5_append(&md5_state, pass_md5 + 16, 5);   /* Add the salt. */
-	md5_finish(&md5_state, pass_md5);
-
-	/* Save the hash in base64-encoded form. */
-	*hash = base64_encode(pass_md5, 21);
+	random_bytes(salt, saltlen);
+	gcry_kdf_derive(password, strlen(password), GCRY_KDF_PBKDF2, GCRY_MD_SHA1,
+			salt, saltlen, rounds, keylen, rawhash);
+	b64salt = base64_encode(salt, saltlen);
+	b64hash = base64_encode(rawhash, keylen);
+
+	*hash = g_strdup_printf("$1$%s$%d$%s", b64salt, rounds, b64hash);
```

Comment:
In absence of a standard, a good example will do, and passlib is as good as any, will change.